File: src/backend/server/data/pipeline.py
from dataclasses import dataclass
from typing import Any, Literal, TypeAlias

from server.data.sources import DataSourceId
from server.data.sources.base import DataSourceDefinition
from server.data.transforms import TransformId
from server.data.transforms.base import TransformDefinition
from server.params import AnyParameter, AnyParameterValue, ParameterValue
from server.util.params import get_params_dict
from server.util.registry import Registry

InstanceId: TypeAlias = int


@dataclass
class SourceConfig:
    type: Literal["source"]
    instance_id: InstanceId
    src_id: DataSourceId
    param_values: list[tuple[str, AnyParameterValue]]

    def get_params(self) -> dict[str, Any]:
        return get_params_dict(self.param_values)


@dataclass
class TransformConfig:
    type: Literal["transform"]
    instance_id: InstanceId
    transform_id: TransformId
    input: InstanceId  # TODO: multiple inputs

    param_values: list[tuple[str, AnyParameterValue]]

    def get_params(self) -> dict[str, Any]:
        return get_params_dict(self.param_values)


PipelineElement: TypeAlias = SourceConfig | TransformConfig


@dataclass
class PipelineConfig:
    elements: list[PipelineElement]
    value_output: InstanceId
    label_output: InstanceId

File: src/frontend/src/lib/types/pipeline.ts
import { type ParameterValue } from "./parameter";
import { type NetworkLayoutDescription } from "./layout";
import type { MetaData } from "./metadata";
import type { TransformConfig } from "./transform";
import type { SourceConfig } from "./source";

export type InstanceId = number;
export type PipelineId = number;

export type PipelineElement = SourceConfig | TransformConfig;

export type PipelineConfig = {
    elements: PipelineElement[];
    value_output: InstanceId;
    label_output: InstanceId;
}

export type PipelineVersion = number;

export type PipelineInfoDescription = {
    version: PipelineVersion
}

export type PipelineMetaDescription = MetaData & {}

export type NetworkPipelineDescription = {
    id: PipelineId,
    content: {
        data: PipelineConfig,
        meta: PipelineMetaDescription,
        layout: NetworkLayoutDescription,
    },
    info?: PipelineInfoDescription
};